### PR TITLE
Add tests to the default interceptor implementations

### LINKED ISSUE #12: Add middleware mechanism
# Description
To be able to be highly flexible we shall add a mechanism where we provide such functionality like logging or setting headers by having a middleware component which has a `process` function which takes the request, manipulates it and finally returns it so other middleware components can `process` it further.

So for example when having a log middleware it will just log the request and return it. 

These middleware components shall be exchangeable and extendable so when having the need to add a special handling one implements a middleware component conforming to the existing components and add it to the middleware stack.

### PR DESCRIPTION
**Pull request by SimonNumberTwo**
This PR adds some tests to the default interceptors, furthermore the `AuthenticationRequestInterceptor` is improved by not setting the authorisation header in case of basic auth, if the UTF-8 encoding is not working. Unfortunately there is no test for having a basic authentication mechanism set and thus having a credentials string which is not UTF-8 encodable.

Closes #12

### CODE REVIEW

File: Sources/Jetworking/Interceptor/Request/AuthenticationRequestInterceptor.swift [hunk 1]
```diff
@@ -39,14 +39,14 @@ public final class AuthenticationRequestInterceptor: RequestInterceptor {
...
             let credentialsString = "\(username):\(password)"
             if let credentialsData = credentialsString.data(using: .utf8) {
                 let base64Credentials = credentialsData.base64EncodedString(options: [])
-                authString = "\(Constants.basicAuthStringPrefix) \(base64Credentials)"
+                let authString = "\(Constants.basicAuthStringPrefix) \(base64Credentials)"
+                return [Constants.authorizationHeaderKey: authString]
             }
 
-            return [Constants.authorizationHeaderKey: authString]
+            return nil
```
Comment (JensK611) [inline Sources/Jetworking/Interceptor/Request/AuthenticationRequestInterceptor.swift:49]: Could we use a guard instead?

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    
+    func testAuthenticationMethodBasicAuthentication() {
+        let authenticationRequestInterceptor: AuthenticationRequestInterceptor = .init(authenticationMethod: .basicAuthentication(username: "username", password: "password"))
+
+        var request: URLRequest = .init(url: URL(string: "https://www.google.com")!)
+        request = authenticationRequestInterceptor.intercept(request)
+
+        if let headerFields = request.allHTTPHeaderFields {
+            print(headerFields)
```
Comment (JensK611) [inline Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:28]: Please remove print

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            XCTAssertTrue((headerFields.contains(where: { $0.key == "Authorization" && $0.value == "Basic dXNlcm5hbWU6cGFzc3dvcmQ=" })))
+        } else {
+            assertionFailure()
```
Comment (JensK611) [inline Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:31]: Please use XCTFail(...)

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    func testAuthenticationMethodBearerToken() {
+        let authenticationRequestInterceptor: AuthenticationRequestInterceptor = .init(authenticationMethod: .bearerToken(token: "token"))
+
+        var request: URLRequest = .init(url: URL(string: "https://www.google.com")!)
+        request = authenticationRequestInterceptor.intercept(request)
+
+        if let headerFields = request.allHTTPHeaderFields {
+            XCTAssertTrue((headerFields.contains(where: { $0.key == "Authorization" && $0.value == "Bearer token" })))
+        } else {
+            assertionFailure()
```
Comment (JensK611) [inline Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:44]: Please use XCTFail(...)

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        XCTAssertTrue(request.allHTTPHeaderFields == nil)
+    }
+    
+    func testAuthenticationMethodBasicAuthentication() {
+        let authenticationRequestInterceptor: AuthenticationRequestInterceptor = .init(authenticationMethod: .basicAuthentication(username: "username", password: "password"))
+
+        var request: URLRequest = .init(url: URL(string: "https://www.google.com")!)
+        request = authenticationRequestInterceptor.intercept(request)
+
+        if let headerFields = request.allHTTPHeaderFields {
```
Comment (JensK611) [inline Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:27]: You could use a guard instead

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            authenticationMethod: .custom(headerKey: "CustomAuthorizationHeaderKey", headerValue: "CustomAuthorizationHeaderValue")
+        )
+
+        var request: URLRequest = .init(url: URL(string: "https://www.google.com")!)
+        request = authenticationRequestInterceptor.intercept(request)
+
+        if let headerFields = request.allHTTPHeaderFields {
+            XCTAssertTrue((headerFields.contains(where: { $0.key == "CustomAuthorizationHeaderKey" && $0.value == "CustomAuthorizationHeaderValue" })))
+        } else {
+            assertionFailure()
```
Comment (JensK611) [inline Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:59]: Please use XCTFail(...)

File: Sources/Jetworking/Interceptor/Request/AuthenticationRequestInterceptor.swift (hunk 1 above)
Comment (SimonNumberTwo) [reply Sources/Jetworking/Interceptor/Request/AuthenticationRequestInterceptor.swift:49]: Done.

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift (hunk 3 above)
Comment (SimonNumberTwo) [reply Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:31]: Done.

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift (hunk 2 above)
Comment (SimonNumberTwo) [reply Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:28]: Done.

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift (hunk 4 above)
Comment (SimonNumberTwo) [reply Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:44]: Done.

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift (hunk 5 above)
Comment (SimonNumberTwo) [reply Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:27]: Done.

File: Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift (hunk 6 above)
Comment (SimonNumberTwo) [reply Tests/JetworkingTests/AuthenticationRequestInterceptorTests.swift:59]: Done.

**Review by JensK611 — APPROVED** [review]
Awesome! 🚀

**Pull request merged**